### PR TITLE
[v7r0]InputData resolution: check InputSandbox separately than InputData

### PR DESCRIPTION
**Pull request by chaen**
The idea is to fail/reschedule jobs with unavailable InputSandbox. https://github.com/DIRACGrid/DIRAC/pull/4471/files was a first attempt, but it expected the sandbox to be on the same site as the data, which is not guaranteed. 

@fstagni please check carefully. I am not sure how it works though: if the inputSandbox is on a banned SE, it should be rescheduled. But if the file literally does not exist, then the job should be failed. That is true for the InputData as well. And I am not quite sure this is the current behaviour ? 

BEGINRELEASENOTES
*WMS
CHANGE: Check LFN InputSandbox separately from InputData

ENDRELEASENOTES

### CODE REVIEW

File: WorkloadManagementSystem/Executor/InputData.py [hunk 1]
```diff
@@ -372,3 +401,44 @@ def __getSiteCandidates(self, okReplicas, vo):
...
+    self.jobLog.verbose('Catalog replicas lookup time', '%.2f seconds ' % (time.time() - startTime))
+    if not result['OK']:
+      self.log.warn(result['Message'])
+      return result
+
+    isDict = result['Value']
+
+    self.jobLog.verbose("REPLICA DICT", isDict)
+
+    result = self.__checkReplicas(isDict, vo)
```
Comment (chrisburr) [inline WorkloadManagementSystem/Executor/InputData.py:439]: Do you want to keep this debug output for when this is unmarked as WIP?
```suggestion
    isDict = result['Value']
    result = self.__checkReplicas(isDict, vo)
```

Comment (chaen) [reply WorkloadManagementSystem/Executor/InputData.py:439]: that's copy paste from the existing production code, I actually have no idea if this is of any use to anyone or not :D

File: WorkloadManagementSystem/Executor/InputData.py [hunk 2]
```diff
@@ -93,16 +93,24 @@ def optimizeJob(self, jid, jobState):
...
+    inputData = result['Value']
+
+    result = self._getInputSandbox(jobState)
+    if not result['OK']:
+      self.jobLog.error("Cannot retrieve input sandbox", result['Message'])
+      return result
+    inputSandbox = result['Value']
+
+    if not inputData and not inputSandbox:
       self.jobLog.notice("No input data. Skipping.")
```
Comment (fstagni) [inline WorkloadManagementSystem/Executor/InputData.py:109]: Maybe just update this log here to 
```suggestion
      self.jobLog.notice("No input data nor LFN input sandboxes. Skipping.")
```

File: WorkloadManagementSystem/Executor/InputData.py [hunk 3]
```diff
@@ -203,6 +224,12 @@ def _resolveInputData(self, jobState, inputData):
     if not result['OK']:
       self.jobLog.error("Failed to check replicas", result['Message'])
       return result
+    okReplicas = result['Value']
+
+    result = self.__getSiteCandidates(okReplicas, vo)
+    if not result['OK']:
+      self.jobLog.error("Failed to check SiteCandidates", result['Message'])
+      return result
```
Comment (fstagni) [inline WorkloadManagementSystem/Executor/InputData.py:232]: Why did you move this check here? (it was in `__checkReplicas`)

File: WorkloadManagementSystem/Executor/InputData.py [hunk 4]
```diff
@@ -372,3 +401,44 @@ def __getSiteCandidates(self, okReplicas, vo):
...
+
+    isDict = result['Value']
+
+    self.jobLog.verbose("REPLICA DICT", isDict)
+
+    result = self.__checkReplicas(isDict, vo)
+
+    if not result['OK']:
+      self.jobLog.error("Failed to check replicas", result['Message'])
+
```
Comment (fstagni) [inline WorkloadManagementSystem/Executor/InputData.py:443]: This whole method is largely a copy of the first lines of the `_resolveInputData` method. Would you like to unify the 2?

**Comment by fstagni** [comment]
If the IS is an LFN and it doesn't exist, we should fail the job.

File: WorkloadManagementSystem/Executor/InputData.py (hunk 3 above)
Comment (chaen) [reply WorkloadManagementSystem/Executor/InputData.py:232]: because I reuse the call `__checkReplicas` for the Sandbox, and there we do not care about the site

**Comment by chaen** [comment]
> If the IS is an LFN and it doesn't exist, we should fail the job.

yes but we should also do that for the input data. is that the case ? As I said, I do not understand this code...

**Comment by fstagni** [comment]
It's simply https://github.com/DIRACGrid/DIRAC/blob/rel-v7r0/WorkloadManagementSystem/Executor/InputData.py#L149 
as return S_ERROR will fail the optimizer and so the job.

The way you coded should be OK: it will fail the optimizer (and so the job) if the IS or the Input Data are not found.

**Comment by chaen** [comment]
yes but the same will happen if the SE is banned https://github.com/DIRACGrid/DIRAC/blob/rel-v7r0/WorkloadManagementSystem/Executor/InputData.py#L251
How is it decided whether a job goes to fail for good, or if it is rescheduled ?

File: WorkloadManagementSystem/Executor/InputData.py (hunk 4 above)
Comment (chaen) [reply WorkloadManagementSystem/Executor/InputData.py:443]: I tried at first, but it makes the code even more intertwined, so I decided to split it this way

File: WorkloadManagementSystem/Executor/InputData.py (hunk 2 above)
Comment (chaen) [reply WorkloadManagementSystem/Executor/InputData.py]: changed, thanks !

**Comment by atsareg** [comment]
I do not think that replicas on banned SEs are taken into account specifically in order to reschedule the job analyzed. InputData optimizer in its present form does not do any rescheduling at all. All bad replicas are filtered away already in the dm.getReplicasForJobs whether it is on a banned SE or not. If to apply the logic that replicas on banned SEs should trigger rescheduling rather than banning, which is correct, one more check is needed for replicas on banned SEs. May be the dm.getReplicasForJobs can be modified to return "Banned SE" as the replica failure reason. Or may be this should be a separate method which will be called if initial check is not OK to check the possibility of rescheduling instead of failing. 

The job should not be rescheduled but rather put "on hold" similarly to the JobScheduling optimizer, I think. But then the question is - what is the max "On hold" time ? It can be unlimited in a sense that each time the job is put on hold for let say 10mins and the number of prolongations is unlimited unlike Reschedule counter. In the end banned SEs are not meant to be banned forever...

**Comment by chaen** [comment]
Thanks for the detailed answer.
So in any case, it's not for this PR :-) So I guess I can remove the WIP, and just keep the behaviour as is ?

**Comment by fstagni** [comment]
> Thanks for the detailed answer.
> So in any case, it's not for this PR :-) So I guess I can remove the WIP, and just keep the behaviour as is ?

Yes!
